File: 08/python/day8.py
import os
from fractions import gcd

lines = list()
with open("day8.txt", "r") as f:
    lines = f.readlines()

map = lines[0].strip()
print(len(map))
print(map)
print("mapstop")
instr = lines[2:]
instr_dict = { r[:3]: (r[7:10], r[12:15]) for r in instr }

#pos = "AAA"
poses = [k for k in instr_dict.keys() if k[2] == "A"]
print(poses)
j = 0
answers = [0,0,0,0,0,0]
import time
start = time.time()
while True:
    """
    if j % 1000000 == 0 and j != 0:
        end = time.time()
        print(end - start)
        exit()
    """
    if j % len(map) == 0:
        for i in range(len(poses)):
            if poses[i][2] == "Z" and answers[i] == 0:
                print(f"i: {i} j: {j}")
                answers[i] = j
        if all(answers):
            break
    """
    if j % len(map) == 0:
        x = 0
        for i in range(len(poses)):
            if poses[i][2] == "Z":
                x = x + 1;
                #print(f"i: {i} j: {j}")
                #answers[i] = j
        if x == len(poses):
            break
    """
    for i in range(len(poses)):
        if answers[i] != 0:
            continue
        if map[j % len(map)] == "L":
            poses[i] = instr_dict[poses[i]][0]
        elif map[j % len(map)] == "R":
            poses[i] = instr_dict[poses[i]][1]
        else:
            print(f"num: j {j % len(map)}")
            print(f"char :{map[j % len(map)]}")
            exit()
                
    j = j + 1

print(poses)
print(j)
print(answers)
m = 1
for a in answers:
    m = m * a

import math
def lcm(a, b):
    return abs(a*b) // math.gcd(a, b)

result = answers[0]
for i in range(1, len(answers)):
   result = lcm(result, answers[i])

print(result)

#print(a)

File: 08/cpp/day8.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <windows.h>
#include <cmath>
#include <iomanip>
#include <chrono>

using namespace std;

int main() {
    fstream f;
    f.open("day8.txt", ios::in);
    string row, map;
    vector<string> rows;
    vector<string> poses;
    unordered_map<string, array<string, 2>> instr; 

    getline(f, map);
    getline(f, row); // empty trash
    while (!f.eof()) {
        getline(f, row);
        rows.push_back(row);
    }
    for (auto &r : rows) {
        string a = r.substr(0,3);
        string b = r.substr(7,3);
        string c = r.substr(12,3);
        instr[a] = {b, c};
        if (a[2] == 'A') {
            poses.push_back(a);
        }

        cout << "Value for " << a << ": " << instr[a][0] << ", " << instr[a][1] << std::endl;
    }

    long long j = 0;
    cout << setprecision(6) << fixed;
    auto start = std::chrono::system_clock::now();
    while (true) {
        if ((j % 10000000 == 0) && (j > 0)) {
            cout << "1 mil" << endl;
            auto now = std::chrono::system_clock::now();
            auto elapsed = now - start;
            auto perc = j / 21366921060721.0;
            auto ant = 21366921060721.0 / j;
            cout << "percentage: " << 100 * perc << "%" << " elapsed: " << std::chrono::duration_cast<std::chrono::seconds>(elapsed).count() << " seconds" << endl;
            cout << "Time remaining: " << std::chrono::duration_cast<std::chrono::hours>(elapsed * ant).count()/24 << " days" << endl;
        }

        if (j % map.length() == 0) {
            int x = 0;
            for (auto &pos : poses) {
                if (pos[2] == 'Z') {
                    x++;
                }
            }
            if (x == poses.size()) {
                break;
            }
        }

        for (int i = 0; i < poses.size(); i++) {
            if (map[j % map.length()] == 'L') {
                poses[i] = instr[poses[i]][0];
            } else if (map[j % map.length()] == 'R') {
                poses[i] = instr[poses[i]][1];
            } else {
                cout << poses[i] << " error" << endl;
                exit(4);
            }
        }
        j++;
    }

    cout << j << endl;

    return 0;
}